pkg/policies/controlplane/components/differentiator.go

```diff
@@ -0,0 +1,147 @@
+package components
+
+import (
+	"time"
+
+	"go.uber.org/fx"
+
+	policylangv1 "github.com/fluxninja/aperture/api/gen/proto/go/aperture/policy/language/v1"
+	"github.com/fluxninja/aperture/pkg/config"
+	"github.com/fluxninja/aperture/pkg/notifiers"
+	"github.com/fluxninja/aperture/pkg/policies/controlplane/iface"
+	"github.com/fluxninja/aperture/pkg/policies/controlplane/runtime"
+)
+
+// Differentiator is a component that calculates rate of change per tick.
+type Differentiator struct {
+	window time.Duration
+	// readings are saved in ring buffer
+	readings  []point
+	oldestIdx int
+	newestIdx int
+	// capacity is calculated from window duration divided by tick interval
+	capacity    int
+	initialized bool
+}
+
+type point struct {
+	reading   runtime.Reading
+	timestamp time.Time
+}
+
+// Name implements runtime.Component.
+func (*Differentiator) Name() string { return "Differentiator" }
+
+// Type implements runtime.Component.
+func (*Differentiator) Type() runtime.ComponentType { return runtime.ComponentTypeSignalProcessor }
+
+// NewDifferentiator creates a differentiator component.
+func NewDifferentiator(diffProto *policylangv1.Differentiator) runtime.Component {
+	diff := &Differentiator{
+		window:      diffProto.Window.AsDuration(),
+		oldestIdx:   0,
+		newestIdx:   0,
+		capacity:    0,
+		initialized: false,
+	}
+	return diff
+}
+
+// NewDifferentiatorAndOptions creates a differentiator component and its fx options.
+func NewDifferentiatorAndOptions(diffProto *policylangv1.Differentiator, _ int, _ iface.Policy) (runtime.Component, fx.Option, error) {
+	return NewDifferentiator(diffProto), fx.Options(), nil
+}
+
+// Execute implements runtime.Component.Execute.
+func (d *Differentiator) Execute(inPortReadings runtime.PortToValue, tickInfo runtime.TickInfo) (runtime.PortToValue, error) {
+	if !d.initialized {
+		d.init(tickInfo)
+	}
+
+	inputVal := inPortReadings.ReadSingleValuePort("input")
+	outputVal := runtime.NewReading(0)
+
+	// add input to readings array
+	d.readings[d.newestIdx] = point{reading: inputVal, timestamp: tickInfo.Timestamp()}
+
+	if d.oldestIdx != d.newestIdx {
+		oldest := d.readings[d.oldestIdx]
+		newest := d.readings[d.newestIdx]
+
+		if !oldest.reading.Valid() {
+			oldest = d.firstValid(d.oldestIdx, true)
+		}
+
+		// find the newest valid reading for extrapolation
+		if !newest.reading.Valid() {
+			found := d.firstValid(d.newestIdx, false)
+			if found.reading.Valid() && oldest.reading.Valid() {
+				extrapolatedValue := d.extrapolate(oldest, found, tickInfo.Timestamp())
+				newest = point{reading: runtime.NewReading(extrapolatedValue), timestamp: tickInfo.Timestamp()}
+			}
+		}
+
+		// calculate the derivative
+		if oldest.reading.Valid() && newest.reading.Valid() {
+			diff := (newest.reading.Value() - oldest.reading.Value()) /
+				float64(newest.timestamp.UnixMilli()-oldest.timestamp.UnixMilli())
```

Comment:
Not supposed to divide by time: ```Differentiator calculates rate of change per tick```

Simply divide by number of ticks (capacity)

Reply:
ok